File: domain/hanzi_candidate_utils.py
"""Utility helpers for Hanzi candidate pipeline."""

from __future__ import annotations

from typing import Sequence

from domain.hanzi_candidate_types import HanziCandidate


def _norm_space(text: str) -> str:
    return " ".join(text.split())


def _split_syllables(jy_norm: str) -> list[str]:
    jy_norm = _norm_space(jy_norm)
    return jy_norm.split() if jy_norm else []


def _coerce_candidates(raw: object, default_source: str) -> list[HanziCandidate]:
    """Normalise diverse candidate return shapes into `HanziCandidate` objects."""
    if raw is None:
        return []

    out: list[HanziCandidate] = []

    if isinstance(raw, list):
        for item in raw:
            if isinstance(item, str) and item.strip():
                out.append(HanziCandidate(item.strip(), default_source, 0.0))
                continue

            if isinstance(item, (tuple, list)) and len(item) >= 1:
                h = item[0]
                if not isinstance(h, str) or not h.strip():
                    continue
                hanzi = h.strip()

                src = default_source
                freq = 0.0

                if len(item) >= 2:
                    if isinstance(item[1], str) and item[1].strip():
                        src = item[1].strip()
                    elif isinstance(item[1], (int, float)):
                        freq = float(item[1])

                if len(item) >= 3 and isinstance(item[2], (int, float)):
                    freq = float(item[2])

                out.append(HanziCandidate(hanzi, src, freq))

    return out


def _dedupe_keep_first(cands: Sequence[HanziCandidate]) -> list[HanziCandidate]:
    seen: set[str] = set()
    out: list[HanziCandidate] = []
    for c in cands:
        if c.hanzi in seen:
            continue
        seen.add(c.hanzi)
        out.append(c)
    return out


def _simple_rank(cands: Sequence[HanziCandidate]) -> list[HanziCandidate]:
    """A stable, conservative ordering."""
    return sorted(
        list(cands),
        key=lambda c: (-float(c.freq or 0.0), str(c.source), c.hanzi),
    )

File: flutter_app/lib/domain/hanzi_candidate_utils.dart
import 'hanzi_candidate_types.dart';

String normSpace(String text) {
  return text.split(RegExp(r'\s+')).where((p) => p.isNotEmpty).join(' ');
}

List<String> splitSyllables(String jyNorm) {
  final normed = normSpace(jyNorm);
  if (normed.isEmpty) {
    return [];
  }
  return normed.split(' ');
}

List<HanziCandidate> coerceCandidates(dynamic raw, String defaultSource) {
  if (raw == null) {
    return [];
  }
  final out = <HanziCandidate>[];
  if (raw is List) {
    for (final item in raw) {
      if (item is String && item.trim().isNotEmpty) {
        out.add(HanziCandidate(item.trim(), defaultSource, 0.0));
        continue;
      }
      if (item is List && item.isNotEmpty) {
        final h = item[0];
        if (h is! String || h.trim().isEmpty) {
          continue;
        }
        final hanzi = h.trim();
        var src = defaultSource;
        var freq = 0.0;

        if (item.length >= 2) {
          final v = item[1];
          if (v is String && v.trim().isNotEmpty) {
            src = v.trim();
          } else if (v is num) {
            freq = v.toDouble();
          }
        }

        if (item.length >= 3 && item[2] is num) {
          freq = (item[2] as num).toDouble();
        }

        out.add(HanziCandidate(hanzi, src, freq));
      }
    }
  }
  return out;
}

List<HanziCandidate> dedupeKeepFirst(List<HanziCandidate> cands) {
  final seen = <String>{};
  final out = <HanziCandidate>[];
  for (final c in cands) {
    if (seen.contains(c.hanzi)) {
      continue;
    }
    seen.add(c.hanzi);
    out.add(c);
  }
  return out;
}

List<HanziCandidate> simpleRank(List<HanziCandidate> cands) {
  final out = List<HanziCandidate>.from(cands);
  out.sort((a, b) {
    final freqCmp = b.freq.compareTo(a.freq);
    if (freqCmp != 0) {
      return freqCmp;
    }
    final srcCmp = a.source.compareTo(b.source);
    if (srcCmp != 0) {
      return srcCmp;
    }
    return a.hanzi.compareTo(b.hanzi);
  });
  return out;
}
